### ISSUE TITLE
模型能力

### ISSUE THREAD

**Issue by gejianqiao** [decisionintelligence/Aurora#2]
你好，感谢你们的贡献，我想要用这个模型来预测美股指数，请问input length还有给出的预测区间为多少合适

**Comment by ccloud0525** [comment]
> 你好，感谢你们的贡献，我想要用这个模型来预测美股指数，请问input length还有给出的预测区间为多少合适

感谢关注，很有意思的任务。用Aurora做预测时，建议按照以下规则确定输入长度：
1. 确定数据的大致周期，将其的1倍或者2倍设置为inference_token_len的长度
2. 将input_length开成inference_token_len的整数倍

例如，考虑预测日收盘价，周期是5，在使用Aurora时，建议把inference_token_len设置成5，input_length设置成10 * 5 = 50。这么做对于有周期的数据集效果比较明显。可参考我们在一些benchmark上的评测脚本的zero-shot参数设置。

**Comment by gejianqiao** [comment]
非常感谢回复，但是就是这个论文中来看是多模态输入，那就是像金融场景的数据要求的话，是要我自己弄出3种模态的输入嘛，比如新闻图像还有历史数据，还是光历史数据就足够了

**Comment by gejianqiao** [comment]
因为我想要能最大程度的发挥模型的潜力

**Comment by ccloud0525** [comment]
可以仅时序模态输入的，readme中有快速zero-shot推理的例子👍



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2026年2月14日(周六) 下午3:49
收件人: ***@***.***&gt;;
抄送: "XJ ***@***.******@***.***&gt;;
主题: Re: [decisionintelligence/Aurora] 模型能力 (Issue #2)


gejianqiao left a comment (decisionintelligence/Aurora#2)
 
非常感谢回复，但是就是这个论文中来看是多模态输入，那就是像金融场景的数据要求的话，是要我自己弄出3种模态的输入嘛，比如新闻图像还有历史数据，还是光历史数据就足够了

**Comment by gejianqiao** [comment]
这个是我的脚本，我为什么感觉预测的不理想，是中间有什么要改嘛？，import argparse
import os

import pandas as pd
import torch
from huggingface_hub import snapshot_download

from aurora.modeling_aurora import AuroraForPrediction


def download_model(repo_id: str, local_dir: str, repository_type: str, token: str | None):
    if not os.path.exists(local_dir):
        os.makedirs(local_dir)
        print(f"Created directory: {local_dir}")

    print(f"Starting download from '{repo_id}' to '{local_dir}'...")
    try:
        download_path = snapshot_download(
            repo_id=repo_id,
            local_dir=local_dir,
            local_dir_use_symlinks=False,
            repo_type=repository_type,
            token=token,
        )
        print(f"\nSuccess! All files downloaded to: {download_path}")
    except Exception as e:
        print(f"\nAn error occurred during download: {e}")
        weight_path = os.path.join(local_dir, "model.safetensors")
        if not os.path.exists(weight_path):
            raise RuntimeError("Model download failed and no local weights were found.") from e
        print(f"Found existing local weights at: {weight_path}. Continue with local model.")


def load_input_window(csv_path: str, end_date: str, input_length: int) -> pd.DataFrame:
    df = pd.read_csv(csv_path)
    df["date"] = pd.to_datetime(df["date"])
    df = df.sort_values("date").reset_index(drop=True)

    window = df[df["date"] <= pd.Timestamp(end_date)].tail(input_length).copy()
    if len(window) < input_length:
        raise ValueError(
            f"Need {input_length} rows on/before {end_date}, but only found {len(window)}."
        )
    return window


def forecast_visa(
    model_path: str,
    input_window: pd.DataFrame,
    forecast_length: int,
    inference_token_len: int,
    num_samples: int,
) -> pd.DataFrame:
    device = "cuda" if torch.cuda.is_available() else "cpu"
    print(f"Using device: {device}")

    model = AuroraForPrediction.from_pretrained(model_path, local_files_only=True).to(device)
    model.eval()

    seqs = torch.tensor(input_window["close"].values, dtype=torch.float32, device=device).unsqueeze(0)

    with torch.no_grad():
        output = model.generate(
            inputs=seqs,
            max_output_length=forecast_length,
            num_samples=num_samples,
            inference_token_len=inference_token_len,
        )

    # output: [batch=1, num_samples, forecast_length]
    samples = output.squeeze(0).cpu()
    mean_pred = samples.mean(dim=0).numpy()
    p10 = torch.quantile(samples, q=0.10, dim=0).numpy()
    p50 = torch.quantile(samples, q=0.50, dim=0).numpy()
    p90 = torch.quantile(samples, q=0.90, dim=0).numpy()

    start = input_window["date"].iloc[-1] + pd.Timedelta(days=1)
    pred_dates = pd.bdate_range(start, periods=forecast_length)

    result = pd.DataFrame(
        {
            "date": pred_dates.strftime("%Y-%m-%d"),
            "pred_close_mean": mean_pred,
            "pred_close_p10": p10,
            "pred_close_p50": p50,
            "pred_close_p90": p90,
        }
    )
    return result


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--repo_id", type=str, default="DecisionIntelligence/Aurora")
    parser.add_argument("--local_dir", type=str, default="./work_dir")
    parser.add_argument("--repository_type", type=str, default="model")
    parser.add_argument("--hf_token", type=str, default=None)
    parser.add_argument("--skip_download", action="store_true")

    parser.add_argument("--csv_path", type=str, default="data/visa/V_close_daily.csv")
    parser.add_argument("--end_date", type=str, default="2026-02-17")
    parser.add_argument("--input_length", type=int, default=100)
    parser.add_argument("--forecast_length", type=int, default=5)
    parser.add_argument("--inference_token_len", type=int, default=5)
    parser.add_argument("--num_samples", type=int, default=100)
    parser.add_argument(
        "--output_csv",
        type=str,
        default="data/visa/V_close_pred_5d_from_2026-02-17_input100_quantiles.csv",
    )
    args = parser.parse_args()

    if not args.skip_download:
        download_model(
            repo_id=args.repo_id,
            local_dir=args.local_dir,
            repository_type=args.repository_type,
            token=args.hf_token,
        )

    window = load_input_window(args.csv_path, args.end_date, args.input_length)
    pred = forecast_visa(
        model_path=args.local_dir,
        input_window=window,
        forecast_length=args.forecast_length,
        inference_token_len=args.inference_token_len,
        num_samples=args.num_samples,
    )

    print(f"input_start={window['date'].iloc[0].strftime('%Y-%m-%d')}")
    print(f"input_end={window['date'].iloc[-1].strftime('%Y-%m-%d')}")
    print(f"input_len={len(window)}")
    print(pred.to_string(index=False))

    os.makedirs(os.path.dirname(args.output_csv), exist_ok=True)
    pred.to_csv(args.output_csv, index=False)
    print(f"saved={args.output_csv}")


if __name__ == "__main__":
    main()
然后输出的内容就是感觉特别单一没有趋势，
      date  pred_close_mean  pred_close_p10  pred_close_p50  pred_close_p90
2026-02-18       338.059479      337.977448      338.065674      338.138458
2026-02-19       338.056366      338.007629      338.051453      338.114288
2026-02-20       338.078583      337.984741      338.084290      338.176788
2026-02-23       338.018036      337.953369      338.010071      338.095428
2026-02-24       338.186798      338.110962      338.184509      338.290436
saved=data/visa/V_close_pred_5d_from_2026-02-17_input100_quantiles.csv，就是是这个模型还不是论文中的完整版还是什么原因

**Comment by gejianqiao** [comment]
hello，不好意思，就是我预测了下，好像效果不好，你来看看是我没有下载论文中的完整版模型还是我这个参数配置的问题，这个我的运行脚本"""Run Aurora zero-shot forecast for VISA close price."""

import argparse
import os

import pandas as pd
import torch
from huggingface_hub import snapshot_download

from aurora.modeling_aurora import AuroraForPrediction


def download_model(repo_id: str, local_dir: str, repository_type: str, token: str | None):
    if not os.path.exists(local_dir):
        os.makedirs(local_dir)
        print(f"Created directory: {local_dir}")

    print(f"Starting download from '{repo_id}' to '{local_dir}'...")
    try:
        download_path = snapshot_download(
            repo_id=repo_id,
            local_dir=local_dir,
            local_dir_use_symlinks=False,
            repo_type=repository_type,
            token=token,
        )
        print(f"\nSuccess! All files downloaded to: {download_path}")
    except Exception as e:
        print(f"\nAn error occurred during download: {e}")
        weight_path = os.path.join(local_dir, "model.safetensors")
        if not os.path.exists(weight_path):
            raise RuntimeError("Model download failed and no local weights were found.") from e
        print(f"Found existing local weights at: {weight_path}. Continue with local model.")


def load_input_window(csv_path: str, end_date: str, input_length: int) -> pd.DataFrame:
    df = pd.read_csv(csv_path)
    df["date"] = pd.to_datetime(df["date"])
    df = df.sort_values("date").reset_index(drop=True)

    window = df[df["date"] <= pd.Timestamp(end_date)].tail(input_length).copy()
    if len(window) < input_length:
        raise ValueError(
            f"Need {input_length} rows on/before {end_date}, but only found {len(window)}."
        )
    return window


def forecast_visa(
    model_path: str,
    input_window: pd.DataFrame,
    forecast_length: int,
    inference_token_len: int,
    num_samples: int,
) -> pd.DataFrame:
    device = "cuda" if torch.cuda.is_available() else "cpu"
    print(f"Using device: {device}")

    model = AuroraForPrediction.from_pretrained(model_path, local_files_only=True).to(device)
    model.eval()

    seqs = torch.tensor(input_window["close"].values, dtype=torch.float32, device=device).unsqueeze(0)

    with torch.no_grad():
        output = model.generate(
            inputs=seqs,
            max_output_length=forecast_length,
            num_samples=num_samples,
            inference_token_len=inference_token_len,
        )

    samples = output.squeeze(0).cpu()
    mean_pred = samples.mean(dim=0).numpy()
    p10 = torch.quantile(samples, q=0.10, dim=0).numpy()
    p50 = torch.quantile(samples, q=0.50, dim=0).numpy()
    p90 = torch.quantile(samples, q=0.90, dim=0).numpy()

    start = input_window["date"].iloc[-1] + pd.Timedelta(days=1)
    pred_dates = pd.bdate_range(start, periods=forecast_length)

    result = pd.DataFrame(
        {
            "date": pred_dates.strftime("%Y-%m-%d"),
            "pred_close_mean": mean_pred,
            "pred_close_p10": p10,
            "pred_close_p50": p50,
            "pred_close_p90": p90,
        }
    )
    return result


def main():
    parser = argparse.ArgumentParser()

    parser.add_argument("--repo_id", type=str, default="DecisionIntelligence/Aurora")
    parser.add_argument("--local_dir", type=str, default="./work_dir")
    parser.add_argument("--repository_type", type=str, default="model")
    parser.add_argument("--hf_token", type=str, default=None)
    parser.add_argument("--skip_download", action="store_true")

    parser.add_argument("--csv_path", type=str, default="data/visa/V_close_daily.csv")
    parser.add_argument("--end_date", type=str, default="2026-02-17")
    parser.add_argument("--input_length", type=int, default=100)
    parser.add_argument("--forecast_length", type=int, default=5)
    parser.add_argument("--inference_token_len", type=int, default=5)
    parser.add_argument("--num_samples", type=int, default=100)
    parser.add_argument(
        "--output_csv",
        type=str,
        default="data/visa/V_close_pred_5d_from_2026-02-17_input100_quantiles.csv",
    )

    args = parser.parse_args()

    if not args.skip_download:
        download_model(
            repo_id=args.repo_id,
            local_dir=args.local_dir,
            repository_type=args.repository_type,
            token=args.hf_token,
        )

    window = load_input_window(args.csv_path, args.end_date, args.input_length)
    pred = forecast_visa(
        model_path=args.local_dir,
        input_window=window,
        forecast_length=args.forecast_length,
        inference_token_len=args.inference_token_len,
        num_samples=args.num_samples,
    )

    print(f"input_start={window['date'].iloc[0].strftime('%Y-%m-%d')}")
    print(f"input_end={window['date'].iloc[-1].strftime('%Y-%m-%d')}")
    print(f"input_len={len(window)}")
    print(pred.to_string(index=False))

    os.makedirs(os.path.dirname(args.output_csv), exist_ok=True)
    pred.to_csv(args.output_csv, index=False)
    print(f"saved={args.output_csv}")


if __name__ == "__main__":
    main()
然后是结果，
      date  pred_close_mean  pred_close_p10  pred_close_p50  pred_close_p90
2026-02-18       338.059479      337.977448      338.065674      338.138458
2026-02-19       338.056366      338.007629      338.051453      338.114288
2026-02-20       338.078583      337.984741      338.084290      338.176788
2026-02-23       338.018036      337.953369      338.010071      338.095428
2026-02-24       338.186798      338.110962      338.184509      338.290436
saved=data/visa/V_close_pred_5d_from_2026-02-17_input100_quantiles.csv，就是没啥趋势感觉。是不是我得调整什么

**Comment by ccloud0525** [comment]
可以试着调整input_length和inference_token_len，inference_token_len决定了测试时一个token的长度是多少，非常影响实际结果，可以遍历看看，观察一下你这类数据适合多长的token_len，同时，input_length尽量开成inference_token_len的整数倍。当然上述方式其实是针对有明显周期的数据，你可以可视化观察一下，如果你的数据实际周期性不明显，基本上完全是趋势的话，你可以inference_token_len就开48，这个是预训练时候采用的设置。如果这样子还不行，可能就是预训练数据并没有涵盖太多强相关的金融数据，以至于没有泛化了，这也没有办法了。

貌似脚本里有个小问题，图中这步操作完后，其实已经把从目标分布的num_samples个采样给压缩成均值了，然后再算分位点是否有些奇怪。这么做貌似意义就是统计均值预测的所有点（forecast_length），的分位点，画出来应该是两根直线。通常做法是否应该是直接对(num_samples, forecast_length)的张量算分位数。不过可能我没领会你的目的，如果说错了，请勿见怪。

<img width="743" height="323" alt="Image" src="https://github.com/user-attachments/assets/71a620c2-064e-4a86-a969-db4b85899935" />
